hydro_deploy/core/src/ssh.rs

```diff
@@ -300,17 +356,33 @@ impl<T: LaunchedSshHost> LaunchedHost for T {
                         Duration::from_secs(1),
                     )
                     .await?;
+
                 let binary_path_string = binary_path.to_str().unwrap();
                 let args_string = args
                     .iter()
                     .map(|s| shell_escape::unix::escape(Cow::from(s)))
                     .fold("".to_string(), |acc, v| format!("{acc} {v}"));
+                let mut command = format!("{binary_path_string}{args_string}");
+                // Launch with perf if specified, also copy local binary to expected place for perf report to work
+                if let Some(perf) = perf.clone() {
+                    let output_file = perf.output_file.to_str().unwrap();
+                    // let local_binary = PathBuf::from(format!(
+                    //     "{output_file}.bins/home/{user}/hydro-{unique_name}"
+                    // ));
+                    // fs::create_dir_all(local_binary.parent().unwrap()).unwrap();
+                    // fs::write(local_binary, &binary.bin_data).unwrap();
+                    // Attach perf to the command
+                    command = format!("perf record -F {freq} --call-graph dwarf,64000 -o {data_file} {cmd}; perf script --symfs=/ -i {data_file} > {out_file}",
```

Comment:
hmm is this inherently assuming bash-like shell that uses `;` and `>`? We could clean this up by doing just `perf record`. And then instead of the logic where we copy down the `out_file` we could instead `exec` `perf_script` and then stream the stdout down through the `AsyncProcess` APIs. Which is kinda nice because then it keeps the result of `perf script` in memory for processing by the flamegraph library.

Reply:
#1365